8371903: HttpClient: improve handling of HTTP/2 GOAWAY frames with error code

ehs208 opened the pull request:
### Problem

  When the HTTP/2 client receives a GOAWAY frame with a non-zero error code, the current implementation discards both the error code and debug data. Users only see generic "Connection closed by peer" errors without any information about why the server terminated the connection.

  ### Solution

  Per [RFC 9113 §5.4.1](https://www.rfc-editor.org/rfc/rfc9113.html#section-5.4.1), a GOAWAY frame with a non-zero error code indicates a connection error requiring immediate closure. This fix:

  1. **Distinguishes** graceful shutdown (NO_ERROR) from connection errors
  2. **Preserves** error code and debug data in exception messages
  3. **Categorizes** streams based on `lastStreamId`:
     - Streams with ID > `lastStreamId`: Marked as unprocessed for automatic retry
     - Streams with ID ≤ `lastStreamId`: Failed with detailed error information

  ### Changes

  **Core Implementation** (`Http2Connection.java`):
  - Modified `handleGoAway()` to check error code and route appropriately
  - Added `handleGoAwayWithError()` method that:
    - Extracts error code and debug data from GOAWAY frame
    - Creates meaningful error messages with error name, hex code, and debug data
    - Properly categorizes streams for retry or failure

  **Test Infrastructure**:
  - `Http2TestServerConnection.sendGoAway(int, int, byte[])`: Supports custom error codes
  - `Http2TestExchangeImpl.getServerConnection()`: Accessor for test handlers
  - `GoAwayWithErrorTest`: Verifies proper error propagation

  ### Example

  **Before:**
  IOException: Connection closed by peer

  **After:**
  IOException: Received GOAWAY with error code Protocol error (0x1): Invalid HEADERS frame

  ### Testing

  - New `GoAwayWithErrorTest` passes
  - Existing HTTP/2 tests unaffected (NO_ERROR path unchanged)
  - Backward compatible (no public API changes)

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- \[ \] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- \[x\] Change must not contain extraneous whitespace
- \[x\] Commit message must refer to an issue

### Error
&nbsp;⚠️ Pull request body is missing required line: `- [x] I confirm that I make this contribution in accordance with the [OpenJDK Interim AI Policy](https://openjdk.org/legal/ai).`

### Issue
 * [JDK-8371903](https://bugs.openjdk.org/browse/JDK-8371903): HttpClient: improve handling of HTTP/2 GOAWAY frames with error code (**Bug** - P4)


### Reviewers
 * [Daniel Jeliński](https://openjdk.org/census#djelinski) (@djelinski - **Reviewer**) 🔄 Re-review required (review applies to [6dc77e1b](https://git.openjdk.org/jdk/pull/28632/files/6dc77e1b266e30aa2e0b4c0bcf0067e06b65df0f))
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**) 🔄 Re-review required (review applies to [6068c1f9](https://git.openjdk.org/jdk/pull/28632/files/6068c1f9710bde3545c91faa5045b3a650bdba38))
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**) 🔄 Re-review required (review applies to [566dfe23](https://git.openjdk.org/jdk/pull/28632/files/566dfe235ae29ae3028c639d55b6c90886c298c5))

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/28632/head:pull/28632` \
`$ git checkout pull/28632`

Update a local copy of the PR: \
`$ git checkout pull/28632` \
`$ git pull https://git.openjdk.org/jdk.git pull/28632/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 28632`

View PR using the GUI difftool: \
`$ git pr show -t 28632`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/28632.diff">https://git.openjdk.org/jdk/pull/28632.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/28632#issuecomment-3606587679)
</details>

ehs208 committed:
8371903: Preserve error code and debug data from HTTP/2 GOAWAY frames

Distinguish GOAWAY with error from graceful shutdown per RFC 9113.
Fail affected streams with meaningful error messages including error
code and debug data instead of generic 'Connection closed' messages.
Add test to verify proper error propagation."

djelinski commented on src/java.net.http/share/classes/jdk/internal/net/http/Http2Connection.java:
```diff
@@ -1419,6 +1430,68 @@ private void handlePeerUnprocessedStreams(final long lastProcessedStream) {
+            } else {
+                // Stream was being processed - fail it with the connection error
+                final IOException error = new IOException(errorMsg);
+                stream.connectionClosing(error);
```
```suggestion
                stream.connectionClosing(cause.getCloseCause());
```

ehs208 committed:
8371903: HTTP/2 client should preserve error information from GOAWAY frames

Extracts error code and debug data from GOAWAY frames with non-zero
error codes and includes them in exception messages. Streams above
lastStreamId are marked as unprocessed for retry; others fail with
the error details.

ehs208 commented:
@djelinski  
Thanks for the feedback! Made all the changes:

- Now using `cause.getCloseCause()` for consistency
- Merged sendGoAway methods as suggested
- Test now verifies unprocessed stream retries with lastStreamId=1

Let me know if anything else needs adjustment!

djelinski commented:
I'd really like to see some retried requests in the test. Please:
- change the requests to POST; idempotent requests like GET might be retried for other reasons, POST is only retried when unprocessed,
- send all 3 requests at once, so that there's at least a chance that they will be in progress when GOAWAY is sent,
- check that exactly one request fails. It doesn't have to be the first one, the client sometimes reorders the requests internally.

ehs208 commented:
@djelinski Updated per your suggestions:
  - Switched to POST requests (only retried when unprocessed)
  - Sending all 3 requests at once to test race conditions
  - Dynamic lastStreamId based on first arriving request
  - Validates exactly 1 failure and 2 successes regardless of order

  Test now correctly verifies unprocessed stream retry behavior.

ehs208 committed:
8371903: HTTP/2 client should preserve error information from GOAWAY frames

Extracts error code and debug data from GOAWAY frames with non-zero
error codes and includes them in exception messages. Streams above
lastProcessedStream are marked as unprocessed for retry; others fail
with the error details.

djelinski reviewed (commented):
Ok, it looks pretty decent now, but the test is failing intermittently, complaining about connection reset or "EOF reached while reading". This happens sometimes when the server closes a socket while the client is writing to it. Can you stabilize the test?

ehs208 committed:
8371903: HTTP/2 client should preserve error information from GOAWAY frames

Extracts error code and debug data from GOAWAY frames with non-zero
error codes and includes them in exception messages. Streams above
lastProcessedStream are marked as unprocessed for retry; others fail
with the error details.

ehs208 commented:
@djelinski Oops, you're right! I moved `goAwaySentLatch.await()` back to its original position and now it works as expected - only 1 connection is created initially, and the other 2 are retries. Tested locally and all passed.

ehs208 committed:
8371903: Move goAwaySentLatch.await() back to original position

Move the await call to after the first request but before the second
and third requests, ensuring only one initial connection is created
and the other requests are properly retried.

djelinski commented:
Right, we only create one connection initially, and requests 2 and 3 are usually sent over the first connection before they are retried on new connections. This can be observed in the server logs as `resetting stream 3 as REFUSED_STREAM`. Ideally the server should not send anything after sending the GOAWAY frame with error, but I think what we have is good enough,

ehs208 commented:
@djelinski Thanks for the explanation. That makes sense.
If there’s an opportunity in the future, I’d be happy to help clean that part up as well.
Appreciate your guidance.

dfuch reviewed (changes requested):
The copyright line in the file GoAwayWithErrorTest.java needs to be fixed.

ehs208 committed:
GoAwayWithErrorTest.java Copyright update

Co-authored-by: Daniel Fuchs <67001856+dfuch@users.noreply.github.com>

ehs208 committed:
8371903: Use Utils.adjustTimeout() in GoAwayWithErrorTest

ehs208 commented:
@dfuch Thank you for the thorough review. I've applied `Utils.adjustTimeout()` to both timeout values as you suggested.

Also, thanks for the tip about `HttpResponse::connectionLabel()` - that's really helpful to know for verifying connection reuse in the future. For this test, I'll keep the current approach since it's already validating the retry behavior correctly, but I'll definitely keep that in mind for future tests.

dfuch reviewed (approved):
Thanks for the updates. LGTM.

ehs208 commented:
> Thanks for the updates. LGTM.

Thank you!

jaikiran commented on src/java.net.http/share/classes/jdk/internal/net/http/Http2Connection.java:
```diff
@@ -1419,6 +1431,58 @@ private void handlePeerUnprocessedStreams(final long lastProcessedStream) {
+                stream.closeAsUnprocessed();
+                numUnprocessed.incrementAndGet();
+            } else {
+                stream.connectionClosing(cause.getCloseCause());
```
Hello @ehs208, I think this should be done a bit differently. Here we are merely constructing a `TerminationCause` and then closing the streams with that termination cause. But that doesn't guarantee that the connection will have been closed with this termination cause. I think what we should do here is merely mark the unprocessed streams with closeAsUnprocessed() and then just call `close(Http2TerminationCause.forH2Error(errorCode, errorMsg);`. The implementation of the `Http2Connection.Terminator.doTerminate()` already has the necessary logic to close all these (processed) streams with the termination cause. So I think this code should look something like:
```java
final byte[] debugData = frame.getDebugData();
final String debugInfo = debugData.length > 0
        ? new String(debugData, UTF_8)
        : "";

...

streams.forEach((id, stream) -> {
    if (id > lastProcessedStream) {
        stream.closeAsUnprocessed();
        numUnprocessed.incrementAndGet();
    }
});
...
// closes the connection as well as the rest of the (processed) streams
close(Http2TerminationCause.forH2Error(errorCode, errorMsg));
```
I haven't tested this myself.

jaikiran replied:
Hello Daniel, line 1483, does guarantee that the connection will actually get closed (if it isn't already).
In the current proposed form in this PR, the potential issue is that the processed streams are being passed a termination cause (the one from goaway frame), however the connection itself might get closed by a different cause, by the time the close on line 1483 is run. So we can end up in a situation where the connection closure cause is something else and the stream termination cause is something else.

jaikiran commented on test/jdk/java/net/httpclient/http2/GoAwayWithErrorTest.java:
```diff
@@ -0,0 +1,193 @@
+                .version(HTTP_2)
+                .build();
+
+        URI uri = URI.create("https://localhost:" + port + "/test");
```
The server constructed in this test (rightly) binds to loopback address. Using `localhost` here can resolve to a different address on some hosts. Please use `URIBuilder` test library class here instead of using `localhost` to construct the URI. Something like:

```java
URI uri = URIBuilder.newBuilder()
        .scheme("https")
        .host(server.getAddress().getAddress())
        .port(server.getAddress().getPort())
        .path("/test")
        .build();
```

ehs208 committed:
8371903: Address code review feedback

djelinski commented:
Occasionally more than one request fails with the GOAWAY exception. I suppose this might be because of a race between the thread that iterates over the list of streams, and another thread adding new streams to the list; the newly added streams may be missed by the iterator, and then are subsequently closed with the connection's termination cause. I only observed the failure 3 times in 2500 repeats.

The exception itself is not visible in the test output; printStackTrace helps a lot in diagnosing test failures, could you add it?

ehs208 committed:
8371903: Add printStackTrace() to GoAwayWithErrorTest

ehs208 commented:
Thank you @djelinski for finding this. I've added `e.printStackTrace()` to make the exception visible in test output.

If I understand correctly, the issue might be a race between `streams.forEach()` in `handleGoAwayWithError()` and other threads adding new streams. If streams 3 and 5 get added while forEach is running (or just after it finishes), they would miss the `closeAsUnprocessed()` call and get closed with the GOAWAY cause instead. Is that the scenario you're thinking of?

The printStackTrace should help confirm what's actually happening. Let me know if there's anything else I should add.

ehs208 committed:
8371903: Use stateLock to synchronize stream iteration and connection closure

ehs208 commented:
@djelinski   
 I've added `stateLock` around the `streams.forEach()` and `close()` in `handleGoAwayWithError()`.                                                                                                                                                                                                                                                            
                                                                                                                                                                                                                                                                                                                                                               
This should prevent the race condition where streams get added to the map while forEach is running. Since `putStream()` already uses the same `stateLock`, any concurrent stream additions will now block until forEach and close() complete. 

The test passes locally, but I can't reproduce the intermittent failure (only happens 3/2500 times). Hopefully this fixes it.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                   
Just to confirm my understanding: the issue was that streams 3 and 5 could be added to the streams map during forEach execution, so they would miss the `closeAsUnprocessed()` call and get closed with the GOAWAY cause instead. By holding stateLock during forEach and close(), any concurrent `putStream()` calls will block until the connection is closed, at which point they'll fail the `isOpen()` check and the streams will automatically use a new connection. Is that correct?

ehs208 commented:
Hi all, Just following up to see if anyone has a chance to take a look at this.  
Thanks.

djelinski commented:
Well with af092399cd5d188fdec4061d33528f6dc0f3a7f0 the `GoAwayWithErrorTest` deadlocks sometimes, with the following stack traces:
Thread 1:
```
              "java.base\/java.util.concurrent.locks.ReentrantLock.lock(ReentrantLock.java:323)",
              "java.net.http\/jdk.internal.net.http.Http2Connection.tryReserveForPoolCheckout(Http2Connection.java:1604)",
              "java.net.http\/jdk.internal.net.http.Http2ClientImpl.getConnectionFor(Http2ClientImpl.java:109)",
              "java.net.http\/jdk.internal.net.http.ExchangeImpl.attemptHttp2Exchange(ExchangeImpl.java:143)",
              "java.net.http\/jdk.internal.net.http.ExchangeImpl.get(ExchangeImpl.java:122)",
              "java.net.http\/jdk.internal.net.http.Exchange.establishExchange(Exchange.java:409)",
              "java.net.http\/jdk.internal.net.http.Exchange.responseAsyncImpl(Exchange.java:624)",
              "java.net.http\/jdk.internal.net.http.Exchange.responseAsync(Exchange.java:447)",
              "java.net.http\/jdk.internal.net.http.MultiExchange.responseAsyncImpl(MultiExchange.java:483)",
              "java.net.http\/jdk.internal.net.http.MultiExchange.lambda$responseAsync0$0(MultiExchange.java:402)",
```
Thread 2:
```
              "java.base\/java.util.concurrent.locks.ReentrantLock.lock(ReentrantLock.java:323)",
              "java.net.http\/jdk.internal.net.http.Http2ClientImpl.removeFromPool(Http2ClientImpl.java:225)",
              "java.net.http\/jdk.internal.net.http.Http2Connection$Terminator.doTerminate(Http2Connection.java:2135)",
              "java.net.http\/jdk.internal.net.http.Http2Connection$Terminator.terminate(Http2Connection.java:2099)",
              "java.net.http\/jdk.internal.net.http.Http2Connection.close(Http2Connection.java:899)",
              "java.net.http\/jdk.internal.net.http.Http2Connection.handleGoAwayWithError(Http2Connection.java:1481)",
              "java.net.http\/jdk.internal.net.http.Http2Connection.handleGoAway(Http2Connection.java:1411)",
```
connectionPoolLock and stateLock are acquired in a different order, leading to the deadlock.

ehs208 committed:
8371903: Move close() outside stateLock to avoid deadlock

ehs208 commented:
@djelinski 
I addressed the deadlock by moving `close(cause)` outside of stateLock, keeping the stream iteration synchronized while avoiding holding the lock during `close()`.

The change works in local testing, but I don’t have a way to validate this under CI conditions. I’m also unsure whether this is the best solution, so I’d appreciate any guidance or alternative suggestions.

dfuch commented:
One suggestion is to store the `lastProcessedStream` in an AtomicLong at the Http2Connection level and use that information in `doTerminate()` to decide whether to mark the stream as unprocessed.

ehs208 commented:
@jaikiran I see @dfuch suggested using AtomicLong to store lastProcessedStream at the connection level. Should I try implementing this approach, or would you prefer to handle it since you're taking over the work?

ehs208 commented:
/touch

ehs208 commented:
/touch

jaikiran commented:
> Would it be OK, if I took over the work on this one?

I've assigned the issue to myself and have started looking into this.

ehs208 commented:
/touch

jaikiran commented:
It's OK to let this close. I have work in progress for which I'll propose a separate PR in the coming days.
